File: e2e-tests/js/bulletin_fetch.js
// Smoldot
// Copyright (C) 2019-2026  Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import { webcrypto } from "node:crypto";
import {
  addChainFromSpec,
  createSmoldotClient,
  report,
  sendRpcAndWait,
} from "./helpers.js";

const ERR_INVALID_PARAMS = -32602;
const ERR_FAIL = -32810;
const ERR_FAIL_RETRY = -32811;
const ERR_FAIL_BACKOFF = -32812;

const relaySpecPath = process.env.RELAY_CHAIN_SPEC;
const bulletinSpecPath = process.env.BULLETIN_CHAIN_SPEC;
const missingCid = process.env.MISSING_CID;
const payloadsJson = process.env.PAYLOADS_JSON;
if (!relaySpecPath || !bulletinSpecPath || !missingCid || !payloadsJson) {
  console.error(
    "Required env vars: RELAY_CHAIN_SPEC, BULLETIN_CHAIN_SPEC, MISSING_CID, PAYLOADS_JSON",
  );
  process.exit(1);
}
const payloads = JSON.parse(payloadsJson);

const client = createSmoldotClient();
let exitCode = 0;
try {
  const relay = await addChainFromSpec(client, relaySpecPath);
  const bulletin = await addChainFromSpec(client, bulletinSpecPath, {
    potentialRelayChains: [relay],
  });

  for (const payload of payloads) {
    try {
      // Given
      const cid = payload.cid;

      // When
      const hex = await bitswapGetWithRetry(bulletin, cid);

      // Then
      const bytes = hexToBytes(hex);
      const sha = await sha256Hex(bytes);
      const ok = bytes.length === payload.size && sha === payload.sha256;
      report(
        `known-${payload.label}`,
        ok,
        ok ? `${bytes.length} bytes` : `size/sha256 mismatch`,
      );
    } catch (err) {
      report(`known-${payload.label}`, false, err.message);
    }
  }

  try {
    // Given
    const cid = missingCid;

    // When
    const hex = await bitswapGetWithRetry(bulletin, cid);

    // Then
    report(
      "missing-not-found",
      false,
      `expected error ${ERR_FAIL}, got success (${hex.length / 2} bytes)`,
    );
  } catch (err) {
    const code = errorCode(err);
    report(
      "missing-not-found",
      code === ERR_FAIL,
      code === ERR_FAIL ? `code ${code}` : `expected ${ERR_FAIL}, got ${code}`,
    );
  }

  try {
    // Given
    const cid = "not-a-cid";

    // When
    await bitswapGetWithRetry(bulletin, cid);

    // Then
    report(
      "missing-invalid-cid",
      false,
      `expected error ${ERR_INVALID_PARAMS}, got success`,
    );
  } catch (err) {
    const code = errorCode(err);
    report(
      "missing-invalid-cid",
      code === ERR_INVALID_PARAMS,
      code === ERR_INVALID_PARAMS
        ? `code ${code}`
        : `expected ${ERR_INVALID_PARAMS}, got ${code}`,
    );
  }

  for (const payload of payloads.filter((p) => !p.on_partial)) {
    try {
      // Given
      const cid = payload.cid;

      // When
      const hex = await bitswapGetWithRetry(bulletin, cid);

      // Then
      const bytes = hexToBytes(hex);
      const sha = await sha256Hex(bytes);
      const ok = bytes.length === payload.size && sha === payload.sha256;
      report(
        `mixed-${payload.label}`,
        ok,
        ok ? `${bytes.length} bytes` : `size/sha256 mismatch`,
      );
    } catch (err) {
      report(`mixed-${payload.label}`, false, err.message);
    }
  }
} catch (err) {
  console.error(`bulletin_fetch error: ${err?.stack || err}`);
  exitCode = 1;
} finally {
  try {
    await client.terminate();
  } catch (_) {}
}

if (exitCode || process.exitCode) {
  process.exit(exitCode || 1);
}

// Retries the transient BlockRequestFailed/Timeout and NoPeers/QueueFull
// errors smoldot returns while its peer set is warming up.
async function bitswapGetWithRetry(chain, cid, totalBudgetMs = 180_000) {
  const deadline = Date.now() + totalBudgetMs;
  let attempt = 0;
  while (true) {
    attempt += 1;
    const remaining = deadline - Date.now();
    if (remaining <= 0) {
      throw new Error(`bitswap_v1_get timed out after ${totalBudgetMs}ms`);
    }
    try {
      return await sendRpcAndWait(chain, "bitswap_v1_get", [cid], Math.min(60_000, remaining));
    } catch (err) {
      const code = errorCode(err);
      if (code === ERR_FAIL_BACKOFF || code === ERR_FAIL_RETRY) {
        const backoff = Math.min(5_000, 500 * 2 ** Math.min(attempt - 1, 3));
        await new Promise((r) => setTimeout(r, backoff));
        continue;
      }
      throw err;
    }
  }
}

function errorCode(err) {
  const m = /"code":(-?\d+)/.exec(err.message ?? "");
  return m ? Number.parseInt(m[1], 10) : null;
}

function hexToBytes(hex) {
  const stripped = hex.startsWith("0x") ? hex.slice(2) : hex;
  if (stripped.length % 2 !== 0) {
    throw new Error(`odd-length hex: ${stripped.length}`);
  }
  const out = new Uint8Array(stripped.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = Number.parseInt(stripped.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

async function sha256Hex(bytes) {
  const digest = await webcrypto.subtle.digest("SHA-256", bytes);
  return [...new Uint8Array(digest)]
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

File: e2e-tests/tests/bulletin_fetch.rs
// Smoldot
// Copyright (C) 2019-2026  Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use anyhow::{anyhow, Result};
use serde::Serialize;
use smoldot_e2e_tests::{
    bulletin, ensure_smoldot_built,
    harness::{
        bulletin_chain_spec, chain_spec_paths, print_dev_mode_invocation, resolve_bundle,
        spawn_with_snapshots,
    },
    prepare_runtime_spec, resolve_base_dir, run_test,
};

#[derive(Serialize)]
struct PayloadJson {
    label: &'static str,
    cid: String,
    sha256: String,
    size: u64,
    on_partial: bool,
}

/// Smoldot fetches every CID in `bulletin::payloads()`, asserts NotFound
/// for an unrelated CID, and exercises mixed-availability peer selection
/// (some CIDs only on the full-snapshot collator).
#[tokio::test(flavor = "multi_thread")]
async fn bulletin_fetch() -> Result<()> {
    env_logger::try_init().ok();

    let chain_spec = bulletin_chain_spec();
    let base_dir = resolve_base_dir()?;

    let snaps = resolve_bundle(&base_dir)?;

    let network = spawn_with_snapshots(&base_dir, &chain_spec, &snaps, &[]).await?;

    let (relay_spec, bulletin_spec) = chain_spec_paths(&network)?;

    // Overwrite the specs' bootNodes with the live TCP + WebRTC multiaddrs so
    // both hosts can connect (Node over TCP, browser over WebRTC). Both
    // collators must be dialable: payloads are split across full/partial.
    let base_dir_str = base_dir.to_str().ok_or_else(|| anyhow!("non-utf8 base dir"))?;
    let relay_spec =
        prepare_runtime_spec(&network, &relay_spec, &["alice", "bob"], base_dir_str, "relay-spec.json")
            .await?;
    let bulletin_spec = prepare_runtime_spec(
        &network,
        &bulletin_spec,
        &["collator-1", "collator-2"],
        base_dir_str,
        "bulletin-spec.json",
    )
    .await?;

    ensure_smoldot_built();

    let payloads_json = serde_json::to_string(
        &bulletin::payloads()
            .iter()
            .map(|p| PayloadJson {
                label: p.label,
                cid: p.predicted_cid(),
                sha256: p.sha256_hex(),
                size: p.size(),
                on_partial: p.on_partial,
            })
            .collect::<Vec<_>>(),
    )?;
    let missing_cid = bulletin::sha256_cid(b"smoldot-bitswap-not-on-chain").to_string();
    let relay_spec = relay_spec
        .to_str()
        .ok_or_else(|| anyhow!("non-utf8 relay spec path"))?;
    let bulletin_spec = bulletin_spec
        .to_str()
        .ok_or_else(|| anyhow!("non-utf8 bulletin spec path"))?;

    let env_pairs = [
        ("RELAY_CHAIN_SPEC", relay_spec),
        ("BULLETIN_CHAIN_SPEC", bulletin_spec),
        ("PAYLOADS_JSON", payloads_json.as_str()),
        ("MISSING_CID", missing_cid.as_str()),
    ];

    // `DEV_MODE=1` skips the JS bitswap suite and keeps the network alive
    // for `KEEP_ALIVE_SECS` seconds (default 36000) so a developer can
    // run the JS client manually (the printed `node …` invocation has
    // every env var the test would have set).
    if std::env::var("DEV_MODE").is_ok() {
        let mut dev_env = env_pairs.to_vec();
        dev_env.push(("TEST_NAME", "bulletin_fetch"));
        print_dev_mode_invocation(&dev_env, "hosts/node/run.js");
        let secs: u64 = std::env::var("KEEP_ALIVE_SECS")
            .ok()
            .and_then(|s| s.parse().ok())
            .unwrap_or(36000);
        eprintln!(
            "DEV_MODE: keeping zombienet alive for {secs}s (set KEEP_ALIVE_SECS to override)..."
        );
        tokio::time::sleep(std::time::Duration::from_secs(secs)).await;
        return Ok(());
    }

    run_test("bulletin_fetch", &env_pairs)
        .await
        .map_err(|e| anyhow!("test failed: {e}"))
}
